### PR TITLE
Implement IBC Hooks

### PR DESCRIPTION
**Pull request by Kayanski**
This PR aims at adding the MsgTransferResponse when sending ICS20 tokens. 
This will be used for callbacks within abstract and this is a good thing to support here

### CODE REVIEW

File: src/bank.rs [hunk 1]
```diff
@@ -309,34 +328,87 @@ impl Module for BankKeeper {
             self.get_balance(&bank_storage, &Addr::unchecked(IBC_LOCK_MODULE_ADDRESS))?;
         let locked_amount = balances.iter().find(|b| b.denom == packet.denom);
 
-        if let Some(locked_amount) = locked_amount {
+        let contract_exec = if let Some(memo) = packet.memo {
+            // We match the memo to the IBC hooks format
+            // If it matches, we create the ibc hook sender. They will be the recipient of the funds and the sender of the contract call
+            let json: Value = serde_json::from_str(&memo)?;
```
Comment (Buckram123) [inline src/bank.rs]:
```suggestion
            let json: serde_json::Result<Value> = serde_json::from_str(&memo);
```
We can't assume any memo is a json

File: src/bank.rs [hunk 2]
```diff
@@ -309,34 +328,87 @@ impl Module for BankKeeper {
...
+            if let Some(wasm) = json.get("wasm") {
+                let contract = wasm["contract"]
+                    .as_str()
+                    .ok_or(anyhow!("Expected contract string"))?
+                    .to_string();
+                let msg = wasm["msg"].clone();
```
Comment (Buckram123) [inline src/bank.rs]: I would suggest you to create structure for deserialization, also it shouldn't error on invalid structure, but skip action instead

File: src/ibc/simple_ibc.rs [hunk 3]
```diff
@@ -855,6 +870,35 @@ impl IbcSimpleModule {
 
         events.push(ack_event);
 
+        // Finally, we check that the packet has a callback associated.
+        // If it does, we execute the callback and push the events
+        // Finally we execute the Packet Callback, without failing on error
+        if let Ok(packet) = from_json::<Ics20Packet>(&packet.data) {
+            let json: Value = serde_json::from_str(&packet.memo.unwrap_or("{}".to_string()))?;
```
Comment (Buckram123) [inline src/ibc/simple_ibc.rs]: We can't assume memo is a json here as well I think

File: src/bank.rs [hunk 4]
```diff
@@ -393,25 +465,80 @@ impl Module for BankKeeper {
...
-pub fn optional_unwrap_ibc_denom(denom: String, expected_channel_id: String) -> String {
-    let split: Vec<_> = denom.splitn(3, '/').collect();
-    if split.len() != 3 {
-        return denom;
-    }
+fn wrap_ibc_denom_query(channel_id: &str, denom: &str) -> String {
+    let denom_path = format!("{channel_id}/{denom}");
+
+    format!("ibc/{}", hex::encode(keccak256(denom_path.as_bytes())))
+}
```
Comment (Buckram123) [inline src/bank.rs]: Great!

File: src/bank.rs [hunk 5]
```diff
@@ -309,34 +328,87 @@ impl Module for BankKeeper {
             self.get_balance(&bank_storage, &Addr::unchecked(IBC_LOCK_MODULE_ADDRESS))?;
         let locked_amount = balances.iter().find(|b| b.denom == packet.denom);
 
-        if let Some(locked_amount) = locked_amount {
+        let contract_exec = if let Some(memo) = packet.memo {
```
Comment (Buckram123) [inline src/bank.rs]: Would it be possible to put into separate file? memo parsing could grow quite heavliy, depending on what we implement in the future

**Review by Buckram123 — CHANGES_REQUESTED** [review]
That's awesome that we can use memo functionality in multi-test environment now, but I don't think it's right to assume every memo is a json and we shouldn't error if one of the fields is incorrectly structured

File: src/bank.rs (hunk 2 above)
Comment (Kayanski) [reply src/bank.rs]: For the first, yes ! For the second however, the ibc hooks should error if there's a wasm attribute and the structure
--> https://github.com/cosmos/ibc-apps/blob/main/modules/ibc-hooks/README.md#ics20-packet-structure

Comment (Buckram123) [reply src/bank.rs]: > For the first, yes ! For the second however, the ibc hooks should error if there's a wasm attribute and the structure --> https://github.com/cosmos/ibc-apps/blob/main/modules/ibc-hooks/README.md#ics20-packet-structure

Ah, nice. If we have "wasm" key we can freely error on invalid structure!

File: src/bank.rs (hunk 5 above)
Comment (Kayanski) [reply src/bank.rs]: Yes !

File: src/ibc/memo/callback.rs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+            serde_json::from_str::<IbcCallbackRequest>(&packet.memo.unwrap_or("{}".to_string()))
+        {
+            if callback_request.src_callback.is_some() {
+                return Ok(());
+            }
+        }
+    }
+
+    bail!("No ics_20_memo callback")
+}
```
Comment (Buckram123) [inline src/ibc/memo/callback.rs:29]: Nice!

File: src/ibc/memo/ibc_hooks.rs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    Ok(None)
+}
+
+pub fn parse_ibc_hooks_callback_memo(
+    api: &dyn Api,
+    packet: &Ics20Packet,
+) -> anyhow::Result<Option<Addr>> {
+    println!("{:?}", packet.memo);
```
Comment (Buckram123) [inline src/ibc/memo/ibc_hooks.rs]: debug line, in case you missed it by accident

**Review by Buckram123 — COMMENTED** [review]
Nice additions

**Pull request merged**